lib/vrl/stdlib/src/map.rs

```diff
@@ -0,0 +1,164 @@
+                    kind: Kind::Integer,
+                },
+                closure::Variable { kind: Kind::all() },
+            ],
+            output: closure::Output::Any,
+        };
+
+        Some(closure::Definition {
+            inputs: vec![object, array],
+        })
+    }
+}
+
+#[derive(Debug, Clone)]
+struct MapFn {
+    value: Box<dyn Expression>,
+    closure: Closure,
+}
+
+impl Expression for MapFn {
+    fn resolve(&self, ctx: &mut Context) -> Result<Value> {
+        // - First focus on objects, ignore arrays
+        // - A closure can resolve if it knows:
+        //   - Closure variable signature/identifiers
+        //   - Content of the object
+        //   - Access to `ctx`
+        //
+        // - Resolving closure should be agnostic. It takes the values you want to assign to
+        //   different closure variables, and it takes an `Fn` to apply to the data.
+
+        let mut object_result: BTreeMap<String, Value> = BTreeMap::default();
+        let mut array_result: Vec<Value> = Vec::default();
+        let mut is_array = false;
+        let value = self.value.resolve(ctx)?;
+
+        let mut map = |_: &Context, output: Output| -> Result<()> {
+            match output {
+                Output::Object { key, value } => {
+                    object_result.insert(key, value);
+                }
+                Output::Array { element } => {
+                    array_result.push(element);
+                    is_array = true;
+                }
+            };
+
+            Ok(())
+        };
+
+        self.closure.resolve(ctx, value, &mut map)?;
+
+        if is_array {
+            Ok(Value::Array(array_result))
+        } else {
+            Ok(Value::Object(object_result))
+        }
+        // Ok(result.into())
+
+        // let result = match self.value.resolve(ctx)? {
+        //     Value::Object(object) => {
+        //         let mut result = HashMap::default();
+
+        //         for (key, value) in object.into_iter() {
+        //             let ident = key.into();
+
+        //             ctx.state_mut().insert_variable(ident, value);
+        //             let v = self.closure.resolve(ctx)?.try_array()?;
+
+        //             let v = closure.resolve_object(ctx, object)?;
+
+        //             ctx.state_mut().remove_variable(&ident);
+
+        //             result.insert(v[0], v[1]);
+        //         }
+
+        //         result.into()
+        //     }
+        //     Value::Array(array) => {
+        //         let mut result = Vec::with_capacity(array.len());
+
+        //         for (index, value) in array.into_iter().enumerate() {
+        //             ctx.state_mut().insert_variable("index".into(), index);
+
+        //             let v = run(index, value)?;
+        //             result.push(v);
+        //         }
+
+        //         result.into()
+        //     }
+        //     _ => unreachable!("expected object or array"),
+        // };
```

Comment:
Should I delete this commented section, or was this more the final state of the code that you were striving toward?